### ISSUE TITLE
tasks/pose/

### ISSUE THREAD

**Issue by Andriw75** [ultralytics/ultralytics#17725]
Nice to meet you community, here I would like to share a project in charge of saving the results of the key points in a sqlite database based on the ultralytics tracking functions, I hope this code can help save time for some people, likewise I hope for feedback and support to create an LTSM model or some other implementation of pytorch or sklear once the predictions are updated.

# Inference and saving
```python
"""
The goal is to capture the results of a list of pose detection models and a list of videos, and save them in a respective sqlite database.
"""
kwars_main = {
    'ls_models': [
        r'yolo11l_pose.pt',
        r'yolo11m_pose.pt',
        r'yolo11n_pose.pt',
        r'yolo11s_pose.pt',
        r'yolo11x_pose.pt',
    ],
    'paths_vids' : [
                        r"001.mp4",
                        r"002.mp4",
                        r"003.mp4",
                        r"004.mp4",
                        r"005.mp4",
                        r"006.mp4",
                    ],
    'path_db' : r"keypoints.db",
    'step_fps' : 4,
}

import cv2 
class VideoIterator:
    """Commonly due to computational power, export of the image size model, etc.;
        when executing the inference in a real environment there is a waiting time, therefore this class
        will allow us to obtain the images through a time interval, to simulate this delay in
        our videos
    """
    def __init__(self, video_path, interval):
        """
        Initializes the video iterator.
        :param video_path: Path to the video file.
        :param interval: Time interval in seconds between frames.
        """
        self.video_path = video_path
        self.interval = interval
        self.cap = cv2.VideoCapture(video_path)
        
        self.fps = int(self.cap.get(cv2.CAP_PROP_FPS))
        self.total_frames = int(self.cap.get(cv2.CAP_PROP_FRAME_COUNT))
        self.frame_interval = self.fps * interval
        self.frame_count = 0
    
    def __iter__(self):
        return self

    def __next__(self):
        """
        Returns the next valid frame according to the interval.
        :return: The next frame or None if there are no more.
        """
        while self.cap.isOpened():
            ret, frame = self.cap.read()
            if not ret:
                self.cap.release()
                return None

            if self.frame_count % self.frame_interval == 0:
                self.frame_count += 1
                return frame

            self.frame_count += 1

        self.cap.release()
        return None

    def __del__(self):
        if self.cap.isOpened():
            self.cap.release()


from collections import deque
import numpy as np
from pathlib import Path
import sqlite3
import pandas as pd

class Tracker:
    """It will help us to keep our tracking record in the detections when they are at the limit
    """
    def __init__(self, path_sv: str, maxlen: int = 30):
        self.path_sv = Path(path_sv)
        
        if not self.path_sv.exists(): # If it does not exist we will create an empty connection
            with sqlite3.connect(self.path_sv) as conn:
                pass     
            
        self.trackers:dict[int,deque[np.ndarray]] = {} # The record of our trackers
        
        self.maxlen = maxlen
        self.n_keypoints = self.maxlen * 17 * 2
        
    def __Initialize_Id(self, _id: int):
        """if the id does not exist, it creates it"""
        if _id not in self.trackers:
            self.trackers[_id] = deque(maxlen=self.maxlen) 
    
    def Append_Data(self, _id: int, tracker_data: np.ndarray,) -> bool:
        """It will add a data to our tracker based on the tracking id and return if it is full"""
        self.__Initialize_Id(_id)
        self.trackers[_id].append(tracker_data)
        return len(self.trackers[_id]) == self.maxlen
    
    def Clear_Data(self, _id: int):
        if _id in self.trackers:
            self.trackers[_id].clear()   
            
    def Save_Data(self, _id:int, w:int,h:int,name_table:str):
        """We will save the data accordingly,
        - keeping the original points in terms of image resolution (if this data is needed).
        - removing the spacing and through the points x1,y1

        Args:
        _id (int): Tracking ID
        w (int): width of the original images
        h (int): height of the original images
        name_table: YOLO model used
        """
        # array with the points normalized with the space
        array = np.array(self.trackers[_id]) 
        
        # data_series = pd.Series(np.concatenate([array.flatten(), [w, h, '0']]),
        #                 index=[f"COL_{i}" for i in range(self.n_keypoints)] + ['width', 'height', 'clasify']) # points[...],clasify
        
        # with sqlite3.connect(self.path_sv) as conn:
        #     data_series.to_frame().T.to_sql(f"{name_table}_ORIGINAL", conn, if_exists='append', index=False)
        
            
        # we remove the space from the array
        x_nonzero = array[:, :, 0] != 0
        y_nonzero = array[:, :, 1] != 0
        x_min = np.min(array[x_nonzero, 0]) - 0.01 if np.any(x_nonzero) else 0
        y_min = np.min(array[y_nonzero, 1]) - 0.01 if np.any(y_nonzero) else 0
        array[:, :, 0] = np.maximum(array[:, :, 0] - x_min, 0)
        array[:, :, 1] = np.maximum(array[:, :, 1] - y_min, 0)
        
        array[:,:,0] *= w
        array[:,:,1] *= h
        
        s1,s2 = array[:,:,0].max() + 1,array[:,:,1].max()+ 1
        
        array[:,:,0] /= s1
        array[:,:,1] /= s2
        
        array = np.array(array.flatten().tolist() + [s1,s2,'0'])
        data_series = pd.Series(array,
                        index=[f"COL_{i}" for i in range(self.n_keypoints)] + ['width', 'height', 'clasify']) # points[...],clasify
        
        with sqlite3.connect(self.path_sv) as conn:
            data_series.to_frame().T.to_sql(f"{name_table}_CROPPED", conn, if_exists='append', index=False)

from ultralytics import YOLO 

for path_model in kwars_main['ls_models']:
    
    path_model = Path(path_model)
    model = YOLO(path_model, task='pose')
    kwars_model = {
        'imgsz': 1920,
        'conf': 0.65,
        'verbose': False,
        'persist': True,
    }

    for path_vid in kwars_main['paths_vids']:
        path_vid = Path(path_vid)
        
        vid = VideoIterator(path_vid,kwars_main['step_fps'])
        track_np = Tracker(kwars_main['path_db'])
        
        for frame in vid:
            
            if frame is None:
                break
            
            resul = model.track(frame,**kwars_model)[0].cpu().numpy()
            
            if resul.boxes.id is None: # If there are no boxes containing id we will continue
                continue
            
            h, w = resul.orig_shape
            
            for idx, data in enumerate(resul.keypoints.xyn):
                                
                if np.sum(np.any(data != 0, axis=1)) <= 8: # if the 9 or more keypoints are not visible then we will continue
                    continue
                
                _id = int(resul.boxes.id[idx])
                is_save = track_np.Append_Data(_id,data)

                if is_save:
                    track_np.Save_Data(_id,w,h,path_model.stem)
                    track_np.Clear_Data(_id)
                    
#ON WRITING TABLES
with sqlite3.connect(kwars_main['path_db']) as conn:
    for path_model in kwars_main['ls_models']:
        path_model = Path(path_model)
        
        df = pd.read_sql_query(f"SELECT * FROM {path_model.stem}_CROPPED",conn)
        
        df.insert(0, 'ID', range(1, len(df) + 1))
        df.set_index('ID', inplace=True)
        df.to_sql(f'{path_model.stem}_CROPPED', conn, if_exists='replace', index=True)
    
        df = pd.read_sql_query(f"SELECT * FROM  {path_model.stem}_ORIGINAL",conn)
        df.insert(0, 'ID', range(1, len(df) + 1))
        df.set_index('ID', inplace=True)
        df.to_sql(f'{path_model.stem}_ORIGINAL', conn, if_exists='replace', index=True)
```
# Plot Keypoints
This other code fragment is responsible for searching for the results in our database and plotting them accordingly, saving them in output.gif

```python
import numpy as np
import sqlite3
import pandas as pd
from PIL import Image
import cv2

skeleton = [
    [16, 14, '#00ff13'],
    [14, 12, '#00ff13'],
    [17, 15, '#00ff13'],
    [15, 13, '#00ff13'],
    [12, 13, '#ff0000'],
    [6, 12, '#ff0000'],
    [7, 13, '#ff0000'],
    [6, 7, '#e4ff00'],
    [6, 8, '#e4ff00'],
    [7, 9, '#e4ff00'],
    [8, 10, '#e4ff00'],
    [9, 11, '#e4ff00'],
    [2, 3, '#00f0ff'],
    [1, 2, '#00f0ff'],
    [1, 3, '#00f0ff'],
    [2, 4, '#00f0ff'],
    [3, 5, '#00f0ff'],
    [4, 6, '#00f0ff'],
    [5, 7, '#00f0ff']
]

def Hex_To_Bgr(hex_color):
    hex_color = hex_color.lstrip('#')
    return tuple(int(hex_color[i:i+2], 16) for i in (4, 2, 0))

def plot_keypoints(kpts:np.ndarray,s1:int,s2:int) -> Image.Image:
    frame = np.zeros((s2, s1, 3), dtype=np.uint8)
    for sk in skeleton:
        pos1 = (int(kpts[sk[0] - 1, 0]), int(kpts[sk[0] - 1, 1]))
        pos2 = (int(kpts[sk[1] - 1, 0]), int(kpts[sk[1] - 1, 1]))

        
        # Check that the points are within the frame limits
        if pos1[0] % s1 == 0 or pos1[1] % s2 == 0 or pos1[0] < 0 or pos1[1] < 0:
            continue
        if pos2[0] % s1 == 0 or pos2[1] % s2 == 0 or pos2[0] < 0 or pos2[1] < 0:
            continue
            
        cv2.line(frame, pos1, pos2, Hex_To_Bgr(sk[2]), 2)
    return Image.fromarray(frame)

def resize_and_pad(image: Image.Image, max_size: tuple[int, int]) -> Image.Image:
    max_width, max_height = max_size
    new_image = Image.new("RGB", (max_width, max_height), color="black")
    new_image.paste(image, (0, 0))
    return new_image

def get_max_size(images: list[Image.Image]) -> tuple[int, int]:
    max_width = max(img.width for img in images)
    max_height = max(img.height for img in images)
    return max_width, max_height

def create_gif(output:str,images: list[Image.Image], duration: int = 200) -> bytes:
    
    max_size = get_max_size(images) #Get the maximum size of each image
    resized_images = [resize_and_pad(img, max_size) for img in images] #Resize with padding accordingly
    
    resized_images[0].save(output, format="GIF", save_all=True, append_images=resized_images[1:], duration=duration, loop=0) #Save the list of images to a gif file

class MNJ_DB:
    def __init__(self,path_db:str,shape_points:tuple=(30,17,2)):
        self.path_db = path_db
        self.shape_points = shape_points
        self.tables ={
            0 : 'yolo11l_pose',
            # 1: 'yolo11m_pose',
        }
        self.sub_tables = {
            0 : 'CROPPED',
            1 : 'ORIGIN'
        }
        
        self.__inicialize_idx()
        
    def __inicialize_idx(self):
        data = {}
        
        with sqlite3.connect(self.path_db) as con:
            
            for md in self.tables.values():
                data[md] = tuple(pd.read_sql_query(f"select ID from {md}_CROPPED",con)['ID'])
                
        self.tp_idx = data
                
    def get_data(self,idx,model:int,sub_table:int=0):
        query_keypoints = f"SELECT * FROM {self.tables[model]}_{self.sub_tables[sub_table]} WHERE id = {idx}"
        
        with sqlite3.connect(self.path_db) as con:
            cursor = con.execute(query_keypoints)
            result = cursor.fetchone()    
            cursor.close()

        if result is None:
            return {
                'data' : None,
                'size' : None,
                'clasify' : None,
            }
        
        return {
            'data' : np.array(result[1:-3]).astype(float).reshape(self.shape_points),
            'size': (int(float(result[-3])), int(float(result[-2]))),  # Convertir a float y luego a int
            'clasify' : int(result[-1])
                }
    
    def update_keypoint(self,_id:int,new_clasify:int,model:int,sub_table:int=0):
        query_update = f"UPDATE {self.tables[model]}_{self.sub_tables[sub_table]} SET clasify={new_clasify} WHERE ID={_id}"
        with sqlite3.connect(self.path_db) as con:
            cursor = con.execute(query_update)
            con.commit()
            if cursor.rowcount > 0:
                return True
            return False
    
def get_gif(manager:MNJ_DB,idx,model:int=0,sub_table:int=0,velocidad:int=200)->dict:
    data = manager.get_data(idx,model,sub_table)
    
    if data['data'] is None:
        return {
            'gif' : None,
            'clasify' : None
        }
    
    
    array = data['data']
    
    array[:,:,0] *= data['size'][0]
    array[:,:,1] *= data['size'][1]
    
    ls = []
            
    for kpts in array:
        ls.append(plot_keypoints(kpts,int(data['size'][0]),int(data['size'][1])))
    
    create_gif("output.gif",ls,velocidad)
```
As I mentioned previously, I would like a way to be able to train a model with pytorch or sklearn once my data has been captured and classified, taking into account that I will provide a matrix (captured_results, 17, 2).

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Andriw75, thank you for sharing your interesting project with the Ultralytics community 🚀! It looks like you are working on an innovative approach to saving pose detection results in a SQLite database and seeking feedback on extending it with an LSTM model or other implementations.

If this is a 🐛 Bug Report, or if you're facing any issues with your code, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us assist you better.

For questions related to custom training or implementing machine learning models like LSTM, please provide detailed information on your current setup and what specific feedback you seek. Be sure to check out our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to ensure you're using the best practices for training.

Do explore the wealth of information available in our [Docs](https://docs.ultralytics.com), especially if you are new to Ultralytics. You can find practical examples in [Python](https://docs.ultralytics.com/usage/python/) and via the [CLI](https://docs.ultralytics.com/usage/cli/) that might be helpful.

Connect with our community for more interactive discussions and assistance:
- Join our [Discord](https://ultralytics.com/discord) 🎧 for real-time support.
- Engage in deeper discussions on our [Discourse](https://community.ultralytics.com) platform.
- Share and learn with fellow enthusiasts on our [Subreddit](https://reddit.com/r/ultralytics).

## Upgrade

To ensure you are working with the latest features and bug fixes, please upgrade to the latest `ultralytics` package with:

```bash
pip install -U ultralytics
```

## Environments

Your project can benefit from running in any of the following environments which are preconfigured for YOLO tasks:

- **Notebooks** with free GPU access: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM, see the [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI, see the [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**, see the [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

The current status of our [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) can guide you on the operational status of the YOLO models across different platforms.

This is an automated response, but rest assured, an Ultralytics engineer will also review your project and provide further assistance soon. Thank you again for contributing to our community! 😊

**Comment by glenn-jocher** [comment]
@Andriw75 thank you for sharing your project! It’s impressive and could greatly benefit others in the community working with pose estimation and data management. For using the captured datasets (`captured_results`) to train a model with PyTorch or sklearn, you can explore the following general approach:

1. **Data Preparation**: Flatten or reshape the pose keypoints (e.g., `(17, 2)` to `(34,)` or normalized representations). Ensure your data is labeled and aligned with the model's input requirements.
   
2. **Training with PyTorch**:
   - Create a simple neural network model.
   - Use your labeled data (`captured_results`) for supervised learning (classification or regression tasks).
   
3. **Training with sklearn**: 
   - Reshape the keypoints data into a format compatible with sklearn models.
   - Train using algorithms like Random Forest, Support Vector Machines, or Decision Trees based on your target problem.

If you'd like further detailed insights or code snippets for this, feel free to ask. Great job on your implementation, and looking forward to seeing more updates!
